crates/notion-core/src/session.rs

```diff
@@ -158,40 +164,162 @@ impl Session {
+    pub fn fetch_npm(&mut self, version_spec: &VersionSpec) -> Fallible<Fetched<PackageVersion>> {
+        let inventory = self.inventory.get_mut()?;
+        let hooks = self.hooks.get()?;
+        inventory
+            .packages
+            .fetch("npm".to_string(), version_spec, hooks.package.as_ref())
+    }
+
+    /// Fetches a Package version matching the specified semantic versioning requirements.
+    pub fn fetch_package(
+        &mut self,
+        name: String,
+        version_spec: &VersionSpec,
+    ) -> Fallible<Fetched<PackageVersion>> {
         let inventory = self.inventory.get_mut()?;
         let hooks = self.hooks.get()?;
-        inventory.fetch(&tool, hooks)
+        inventory
+            .packages
+            .fetch(name, version_spec, hooks.package.as_ref())
     }
 
-    /// Updates toolchain in package.json with the Tool version matching the specified semantic
+    /// Updates toolchain in package.json with the Node version matching the specified semantic
     /// versioning requirements.
-    pub fn pin(&mut self, toolspec: &ToolSpec) -> Fallible<()> {
+    pub fn pin_node(&mut self, version_spec: &VersionSpec) -> Fallible<()> {
         if let Some(ref project) = self.project()? {
-            let distro_version = self.fetch(toolspec)?.into_version();
-            project.pin(&distro_version)?;
+            let node_version = self.fetch_node(version_spec)?.into_version();
+            project.pin_node(&node_version)?;
         } else {
             throw!(ErrorDetails::NotInPackage);
         }
         Ok(())
     }
 
+    /// Updates toolchain in package.json with the Yarn version matching the specified semantic
+    /// versioning requirements.
+    pub fn pin_yarn(&mut self, version_spec: &VersionSpec) -> Fallible<()> {
+        if let Some(ref project) = self.project()? {
+            let yarn_version = self.fetch_yarn(version_spec)?.into_version();
+            project.pin_yarn(&yarn_version)?;
+        } else {
+            throw!(ErrorDetails::NotInPackage);
+        }
+        Ok(())
+    }
+
+    /// Updates toolchain in package.json with the Npm version matching the specified semantic
+    /// versioning requirements.
+    pub fn pin_npm(&mut self, version_spec: &VersionSpec) -> Fallible<()> {
+        if let Some(ref project) = self.project()? {
+            let inventory = self.inventory.get_mut()?;
+            let hooks = self.hooks.get()?;
+            let npm_version = inventory
+                .packages
+                .resolve("npm".to_string(), version_spec, hooks.package.as_ref())?
+                .version;
+            project.pin_npm(&npm_version)?;
+        } else {
+            throw!(ErrorDetails::NotInPackage);
+        }
+        Ok(())
+    }
+
+    /// Gets the installed UserTool with the input name, if any.
+    pub fn get_user_tool(&mut self, tool_name: &OsStr) -> Fallible<Option<UserTool>> {
```

Comment:
Maybe add a comment referencing the toolchain RFC to provide a definition of "user tool" here and/or where the `UserTool` type is defined.

Reply:
done!